File: a-a-d-an/src/app/core/services/cart.service.ts
import { Injectable, signal, computed } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable, tap } from 'rxjs';
import { Order, CheckoutRequest } from '../../shared/models/order.model';

export interface CartItem {
  cartItemId: string;
  productId: string;
  productName: string;
  primaryImageUrl: string;
  unitPrice: number;
  quantity: number;
  subtotal: number;
}

export interface CartResponse {
  cartId: string;
  items: CartItem[];
  totalAmount: number;
  totalItems: number;
}

@Injectable({ providedIn: 'root' })
export class CartService {
  private readonly CART_API = 'http://localhost:8080/api/cart';
  private readonly CHECKOUT_API = 'http://localhost:8080/api/checkout';

  cart = signal<CartResponse>({ cartId: '', items: [], totalAmount: 0, totalItems: 0 });
  items = computed(() => this.cart().items);
  totalPrice = computed(() => this.cart().totalAmount);
  totalItems = computed(() => this.cart().totalItems);

  constructor(private http: HttpClient) {}

  loadCart(): Observable<CartResponse> {
    return this.http.get<CartResponse>(this.CART_API).pipe(
      tap(cart => this.cart.set(cart))
    );
  }

  addItem(productId: string, quantity = 1): Observable<CartResponse> {
    return this.http.post<CartResponse>(`${this.CART_API}/items`, { productId, quantity }).pipe(
      tap(cart => this.cart.set(cart))
    );
  }

  removeItem(productId: string): Observable<CartResponse> {
    return this.http.delete<CartResponse>(`${this.CART_API}/items/${productId}`).pipe(
      tap(cart => this.cart.set(cart))
    );
  }

  clearCart(): Observable<void> {
    return this.http.delete<void>(this.CART_API);
  }

  checkout(request: CheckoutRequest): Observable<Order> {
    return this.http.post<Order>(this.CHECKOUT_API, request);
  }
}

File: a-a-d-sb/src/main/java/com/ecommerce/analytics/service/CartService.java
package com.ecommerce.analytics.service;

import com.ecommerce.analytics.dto.request.CartItemRequest;
import com.ecommerce.analytics.dto.response.CartItemResponse;
import com.ecommerce.analytics.dto.response.CartResponse;
import com.ecommerce.analytics.exception.ResourceNotFoundException;
import com.ecommerce.analytics.model.Cart;
import com.ecommerce.analytics.model.CartItem;
import com.ecommerce.analytics.model.Product;
import com.ecommerce.analytics.model.User;
import com.ecommerce.analytics.repository.*;
import jakarta.persistence.EntityManager;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.math.BigDecimal;
import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
public class CartService {

    private final CartRepository cartRepository;
    private final CartItemRepository cartItemRepository;
    private final ProductRepository productRepository;
    private final UserRepository userRepository;
    private final ProductImageRepository productImageRepository;
    private final EntityManager entityManager;

    @Transactional(readOnly = true)
    public CartResponse getCart(String userId) {
        Cart cart = getOrCreateCart(userId);
        return toCartResponse(cart);
    }

    @Transactional
    public CartResponse addOrUpdateItem(String userId, CartItemRequest request) {
        Cart cart = getOrCreateCart(userId);
        Product product = productRepository.findByProductIdAndIsActiveTrue(request.getProductId())
                .orElseThrow(() -> new ResourceNotFoundException("Product not found: " + request.getProductId()));

        Optional<CartItem> existing = cartItemRepository.findByCartCartIdAndProductProductId(cart.getCartId(), request.getProductId());
        if (existing.isPresent()) {
            CartItem item = existing.get();
            item.setQuantity(request.getQuantity());
            cartItemRepository.save(item);
        } else {
            CartItem newItem = CartItem.builder()
                    .cart(cart)
                    .product(product)
                    .quantity(request.getQuantity())
                    .build();
            cartItemRepository.save(newItem);
        }
        entityManager.flush();
        entityManager.refresh(cart);
        return toCartResponse(cart);
    }

    @Transactional
    public CartResponse removeItem(String userId, String productId) {
        Cart cart = cartRepository.findByUserUserId(userId)
                .orElseThrow(() -> new ResourceNotFoundException("Cart not found"));
        cartItemRepository.findByCartCartIdAndProductProductId(cart.getCartId(), productId)
                .ifPresent(cartItemRepository::delete);
        entityManager.flush();
        entityManager.refresh(cart);
        return toCartResponse(cart);
    }

    @Transactional
    public void clearCart(String userId) {
        cartRepository.findByUserUserId(userId).ifPresent(cart -> {
            cart.getItems().clear();
            cartRepository.save(cart);
        });
    }

    private Cart getOrCreateCart(String userId) {
        return cartRepository.findByUserUserId(userId).orElseGet(() -> {
            User user = userRepository.findById(userId)
                    .orElseThrow(() -> new ResourceNotFoundException("User not found: " + userId));
            return cartRepository.save(Cart.builder().user(user).build());
        });
    }

    private CartResponse toCartResponse(Cart cart) {
        List<CartItemResponse> items = cart.getItems().stream().map(item -> {
            String imageUrl = productImageRepository
                    .findByProductProductIdAndIsPrimaryTrue(item.getProduct().getProductId())
                    .map(img -> img.getImageUrl()).orElse(null);
            BigDecimal subtotal = item.getProduct().getBasePrice()
                    .multiply(BigDecimal.valueOf(item.getQuantity()));
            return CartItemResponse.builder()
                    .cartItemId(item.getCartItemId())
                    .productId(item.getProduct().getProductId())
                    .productName(item.getProduct().getName())
                    .primaryImageUrl(imageUrl)
                    .unitPrice(item.getProduct().getBasePrice())
                    .quantity(item.getQuantity())
                    .subtotal(subtotal)
                    .build();
        }).collect(Collectors.toList());

        BigDecimal total = items.stream()
                .map(CartItemResponse::getSubtotal)
                .reduce(BigDecimal.ZERO, BigDecimal::add);

        return CartResponse.builder()
                .cartId(cart.getCartId())
                .items(items)
                .totalAmount(total)
                .totalItems(items.stream().mapToInt(CartItemResponse::getQuantity).sum())
                .build();
    }
}
